File: projects/order-creation-queries/app/src/test/java/com/robinfood/app/usecases/getordersdaily/GetOrdersDailyUseCaseTest.java
package com.robinfood.app.usecases.getordersdaily;

import com.robinfood.app.mocks.BrandDTOMock;
import com.robinfood.app.mocks.MenuBrandDTOMock;
import com.robinfood.app.mocks.OrderDailyDTOMock;
import com.robinfood.app.mocks.StoreDTOMock;
import com.robinfood.app.usecases.gettokenbusinesscapability.GetTokenBusinessCapabilityUseCase;
import com.robinfood.core.enums.Result;
import com.robinfood.core.models.domain.TokenModel;
import com.robinfood.repository.configuration.store.IStoreRepository;
import com.robinfood.repository.menubrandstore.MenuBrandStoreRepository;
import com.robinfood.repository.orderdaily.OrderDailyRepository;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.ArgumentMatchers.anyLong;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class GetOrdersDailyUseCaseTest {

    @Mock
    private GetTokenBusinessCapabilityUseCase getTokenBusinessCapabilityUseCase;

    @Mock
    private MenuBrandStoreRepository menuBrandStoreRepository;

    @Mock
    private OrderDailyRepository orderDailyRepository;

    @Mock
    private IStoreRepository storeRepository;

    @InjectMocks
    private GetOrdersDailyUseCase getOrdersDailyUseCase;

    @Test
    void test_Get_Orders_Daily_Returns_Correctly() {

        final String token = "token";

        when(getTokenBusinessCapabilityUseCase.invoke()).thenReturn(
                TokenModel.builder()
                        .accessToken(token)
                        .build()
        );

        when(orderDailyRepository.getOrderDaily(27L, "time", token))
                .thenReturn(new Result.Success<>(List.of(OrderDailyDTOMock.getDataDefault())));

        when(menuBrandStoreRepository.getMenuBrandStore(27L, token))
                .thenReturn(new Result.Success<>(List.of(MenuBrandDTOMock.getDataDefault())));

        when(storeRepository.getStore(27L, token))
                .thenReturn(new Result.Success<>(StoreDTOMock.getDataDefault()));

        getOrdersDailyUseCase.invoke(27L, "time");

        verify(getTokenBusinessCapabilityUseCase).invoke();
        verify(orderDailyRepository).getOrderDaily(anyLong(), anyString(), anyString());
        verify(menuBrandStoreRepository).getMenuBrandStore(anyLong(), anyString());
    }

    @Test
    void test_Get_Orders_Daily_Returns_Correctly_Not_Found_Brand() {

        final String token = "token";

        when(getTokenBusinessCapabilityUseCase.invoke()).thenReturn(
                TokenModel.builder()
                        .accessToken(token)
                        .build()
        );

        when(orderDailyRepository.getOrderDaily(27L, "time", token))
                .thenReturn(new Result.Success<>(List.of(OrderDailyDTOMock.getDataDefaultNotFoundBrand())));

        when(menuBrandStoreRepository.getMenuBrandStore(27L, token))
                .thenReturn(new Result.Success<>(List.of(MenuBrandDTOMock.getDataDefault())));

        when(storeRepository.getStore(27L, token))
                .thenReturn(new Result.Success<>(StoreDTOMock.getDataDefault()));

        getOrdersDailyUseCase.invoke(27L, "time");

        verify(getTokenBusinessCapabilityUseCase).invoke();
        verify(orderDailyRepository).getOrderDaily(anyLong(), anyString(), anyString());
        verify(menuBrandStoreRepository).getMenuBrandStore(anyLong(), anyString());
    }

    @Test
    void test_Get_Orders_Daily_Returns_Correctly_Error_Order_BC() {

        final String token = "token";

        when(getTokenBusinessCapabilityUseCase.invoke()).thenReturn(
                TokenModel.builder()
                        .accessToken(token)
                        .build()
        );

        when(orderDailyRepository.getOrderDaily(27L, "time", token))
                .thenReturn(new Result.Error(new Exception("Some error"), HttpStatus.BAD_REQUEST));

        assertThrows(ResponseStatusException.class, () -> {
            getOrdersDailyUseCase.invoke(27L, "time");
        });
    }

    @Test
    void test_Get_Orders_Daily_Returns_Correctly_Error_Menu_BC() {

        final String token = "token";

        when(getTokenBusinessCapabilityUseCase.invoke()).thenReturn(
                TokenModel.builder()
                        .accessToken(token)
                        .build()
        );

        when(orderDailyRepository.getOrderDaily(27L, "time", token))
                .thenReturn(new Result.Success<>(List.of(OrderDailyDTOMock.getDataDefault())));

        when(menuBrandStoreRepository.getMenuBrandStore(27L, token))
                .thenReturn(new Result.Error(new Exception("Some error"), HttpStatus.BAD_REQUEST));

        assertThrows(ResponseStatusException.class, () -> {
            getOrdersDailyUseCase.invoke(27L, "time");
        });
    }

    @Test
    void test_Get_Store_Returns_Correctly_Error_Menu_BC() {

        final String token = "token";

        when(getTokenBusinessCapabilityUseCase.invoke()).thenReturn(
                TokenModel.builder()
                        .accessToken(token)
                        .build()
        );

        when(orderDailyRepository.getOrderDaily(27L, "time", token))
                .thenReturn(new Result.Success<>(List.of(OrderDailyDTOMock.getDataDefault())));

        when(menuBrandStoreRepository.getMenuBrandStore(27L, token))
                .thenReturn(new Result.Success<>(List.of(MenuBrandDTOMock.getDataDefault())));

        when(storeRepository.getStore(27L, token))
                .thenReturn(new Result.Error(new Exception("Some error"), HttpStatus.BAD_REQUEST));

        assertThrows(ResponseStatusException.class, () -> {
            getOrdersDailyUseCase.invoke(27L, "time");
        });
    }
}

File: projects/posv2/app/src/test/kotlin/com/robinfood/app/usecases/orderdaily/GetOrdersDailyUseCaseTest.kt
package com.robinfood.app.usecases.orderdaily

import com.robinfood.app.mocks.OrderDailyDTOsMocks
import com.robinfood.core.dtos.OrderDailyResponseDTO
import com.robinfood.core.enums.Result
import com.robinfood.repository.orderdaily.IOrdersDailyRepository
import kotlinx.coroutines.runBlocking
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension
import org.springframework.http.HttpStatus
import org.springframework.web.server.ResponseStatusException

@ExtendWith(MockitoExtension::class)
class GetOrdersDailyUseCaseTest {

    @Mock
    private lateinit var ordersDailyRepository: IOrdersDailyRepository

    @InjectMocks
    private lateinit var getOrdersDailyUseCase: GetOrdersDailyUseCase

    private val orderDailyDTOsMocks = OrderDailyDTOsMocks()

    private val token = "token"
    private val timeZone = "America/Bogota"
    private val storeId = 1L
    private val orderDailyDTO = listOf(orderDailyDTOsMocks.orderDailyDTO)
    private val orderDailyDTOResult =OrderDailyResponseDTO(listOf(orderDailyDTOsMocks.orderDailyDTOResult))
    private val customException = Exception("Some error")
    private val customHttpStatus = HttpStatus.BAD_REQUEST

    @Test
    fun `test that get orders daily use case returns correctly`() {
        runBlocking {
            Mockito.`when`(ordersDailyRepository.getAllOrdersDaily(
                token,
                timeZone,
                storeId
            )).thenReturn(Result.Success(orderDailyDTO))

            val result = getOrdersDailyUseCase.invoke(
                token,
                timeZone,
                storeId
            )

            Assertions.assertEquals(Result.Success(orderDailyDTOResult), result)
        }
    }

    @Test
    fun `test that get orders daily use case returns error`() {
        runBlocking {

            Mockito.`when`(ordersDailyRepository.getAllOrdersDaily(
                    token,
                    timeZone,
                    storeId
            )).thenReturn(Result.Error(customException, customHttpStatus))

            val result = getOrdersDailyUseCase.invoke(
                    token,
                    timeZone,
                    storeId
            )

            Assertions.assertTrue(result is Result.Error)
        }
    }
}
